### PR TITLE
quincy: qa/workunits/windows: backport rbd-wnbd tests

### PR DESCRIPTION
**Pull request by petrutlucian94**
The quincy branch doesn't currently include the rbd-wnbd python tests. For this reason, the ceph-build scripts are fetching a certain version from the main branch and then run each test individually [1].

One issue is that the fio test is failing on quincy and we aren't getting any information. It seemed to be a time out, but it still fails even after the timeout was increased.

By backporting the tests to quincy, the tests will be invoked through the "run-tests.ps1" script, hopefully receiving more information.

This will also allow us to simplify the ceph-build script.

[1] https://github.com/ceph/ceph-build/blob/b5adddec2cb03059d0bc0b5d4ffa7aac7a248fc0/scripts/ceph-windows/run_tests#L70-L85

Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by petrutlucian94** [comment]
@idryomov it failed but at least we now have an error:

```
CommandFailed: Command failed: ('fio', '--thread', '--output-format=json', '--randrepeat=1', '--direct=1', '--gtod_reduce=1', '--name=test', '--bs=2M', '--iodepth=64', '--size=1024M', '--readwrite=rw', '--numjobs=1', '--filename=\\\\.\\PhysicalDrive1', '--verify=crc32c'). Return code: 1. Stdout: b'fio: pid=4800, err=5/file:io_u.c:1841, func=io_u error, error=Input/output error\r\n{\r\n  "fio version" : "fio-3.27",\r\n  "timestamp" : 1674723904,\r\n  "timestamp_ms" : 1674723904971,\r\n  "time" : "Thu Jan 26 09:05:04 2023",\r\n  "global options" : {\r\n    "randrepeat" : "1",\r\n    "direct" : "1",\r\n    "gtod_reduce" : "1"\r\n  },\r\n  "jobs" : [\r\n    {\r\n      "jobname" : "test",\r\n      "groupid" : 0,\r\n      "error" : 5,\r\n      "eta" : 0,\r\n      "elapsed" : 1385,\r\n      "job options" : {\r\n        "name" : "test",\r\n        "bs" : "2M",\r\n        "iodepth" : "64",\r\n        "size" : "1024M",\r\n        "rw" : "rw",\r\n        "numjobs" : "1",\r\n        "filename" : "\\\\\\\\.\\\\PhysicalDrive1",\r\n        "verify" : "crc32c"\r\n      },\r\n      "read" : {\r\n        "io_bytes" : 499122176,\r\n        "io_kbytes" : 487424,\r\n        "bw_bytes" : 636827,\r\n        "bw" : 621,\r\n        "iops" : 0.308767,\r\n        "runtime" : 783763,\r\n        "total_ios" : 242,\r\n        "short_ios" : 0,\r\n        "drop_ios" : 0,\r\n        "slat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "clat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "lat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "bw_min" : 12288,\r\n        "bw_max" : 529033,\r\n        "bw_agg" : 100.000000,\r\n        "bw_mean" : 288926.000000,\r\n        "bw_dev" : 260302.197891,\r\n        "bw_samples" : 3,\r\n        "iops_min" : 6,\r\n        "iops_max" : 258,\r\n        "iops_mean" : 140.666667,\r\n        "iops_stddev" : 126.891029,\r\n        "iops_samples" : 3\r\n      },\r\n      "write" : {\r\n        "io_bytes" : 555745280,\r\n        "io_kbytes" : 542720,\r\n        "bw_bytes" : 709073,\r\n        "bw" : 692,\r\n        "iops" : 0.344492,\r\n        "runtime" : 783763,\r\n        "total_ios" : 270,\r\n        "short_ios" : 0,\r\n        "drop_ios" : 0,\r\n        "slat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "clat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "lat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "bw_min" : 49152,\r\n        "bw_max" : 496970,\r\n        "bw_agg" : 100.000000,\r\n        "bw_mean" : 314340.333333,\r\n        "bw_dev" : 235047.208795,\r\n        "bw_samples" : 3,\r\n        "iops_min" : 24,\r\n        "iops_max" : 242,\r\n        "iops_mean" : 153.000000,\r\n        "iops_stddev" : 114.372199,\r\n        "iops_samples" : 3\r\n      },\r\n      "trim" : {\r\n        "io_bytes" : 0,\r\n        "io_kbytes" : 0,\r\n        "bw_bytes" : 0,\r\n        "bw" : 0,\r\n        "iops" : 0.000000,\r\n        "runtime" : 0,\r\n        "total_ios" : 0,\r\n        "short_ios" : 0,\r\n        "drop_ios" : 0,\r\n        "slat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "clat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "lat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        },\r\n        "bw_min" : 0,\r\n        "bw_max" : 0,\r\n        "bw_agg" : 0.000000,\r\n        "bw_mean" : 0.000000,\r\n        "bw_dev" : 0.000000,\r\n        "bw_samples" : 0,\r\n        "iops_min" : 0,\r\n        "iops_max" : 0,\r\n        "iops_mean" : 0.000000,\r\n        "iops_stddev" : 0.000000,\r\n        "iops_samples" : 0\r\n      },\r\n      "sync" : {\r\n        "total_ios" : 0,\r\n        "lat_ns" : {\r\n          "min" : 0,\r\n          "max" : 0,\r\n          "mean" : 0.000000,\r\n          "stddev" : 0.000000,\r\n          "N" : 0\r\n        }\r\n      },\r\n      "job_runtime" : 783762,\r\n      "usr_cpu" : 0.000000,\r\n      "sys_cpu" : 0.000000,\r\n      "ctx" : 0,\r\n      "majf" : 0,\r\n      "minf" : 0,\r\n      "iodepth_level" : {\r\n        "1" : 0.195312,\r\n        "2" : 0.390625,\r\n        "4" : 0.781250,\r\n        "8" : 1.562500,\r\n        "16" : 3.125000,\r\n        "32" : 73.632812,\r\n        ">=64" : 20.312500\r\n      },\r\n      "iodepth_submit" : {\r\n        "0" : 0.000000,\r\n        "4" : 100.000000,\r\n        "8" : 0.000000,\r\n        "16" : 0.000000,\r\n        "32" : 0.000000,\r\n        "64" : 0.000000,\r\n        ">=64" : 0.000000\r\n      },\r\n      "iodepth_complete" : {\r\n        "0" : 0.000000,\r\n        "4" : 94.813614,\r\n        "8" : 1.620746,\r\n        "16" : 2.917342,\r\n        "32" : 0.486224,\r\n        "64" : 0.162075,\r\n        ">=64" : 0.000000\r\n      },\r\n      "latency_ns" : {\r\n        "2" : 0.000000,\r\n        "4" : 0.000000,\r\n        "10" : 0.000000,\r\n        "20" : 0.000000,\r\n        "50" : 0.000000,\r\n        "100" : 0.000000,\r\n        "250" : 0.000000,\r\n        "500" : 0.000000,\r\n        "750" : 0.000000,\r\n        "1000" : 0.000000\r\n      },\r\n      "latency_us" : {\r\n        "2" : 0.000000,\r\n        "4" : 0.000000,\r\n        "10" : 0.000000,\r\n        "20" : 0.000000,\r\n        "50" : 0.000000,\r\n        "100" : 0.000000,\r\n        "250" : 0.000000,\r\n        "500" : 0.000000,\r\n        "750" : 0.000000,\r\n        "1000" : 0.000000\r\n      },\r\n      "latency_ms" : {\r\n        "2" : 0.000000,\r\n        "4" : 0.000000,\r\n        "10" : 0.000000,\r\n        "20" : 0.000000,\r\n        "50" : 0.000000,\r\n        "100" : 0.000000,\r\n        "250" : 0.000000,\r\n        "500" : 0.000000,\r\n        "750" : 0.000000,\r\n        "1000" : 0.000000,\r\n        "2000" : 0.000000,\r\n        ">=2000" : 0.000000\r\n      },\r\n      "latency_depth" : 64,\r\n      "latency_target" : 0,\r\n      "latency_percentile" : 100.000000,\r\n      "latency_window" : 0\r\n    }\r\n  ]\r\n}\r\n'. Stderr: b'fio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: windows error 121 not handled\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: read offset=457179136, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: read offset=459276288, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: write offset=528482304, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: write offset=513802240, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: write offset=488636416, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: write offset=511705088, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: read offset=486539264, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: read offset=488636416, buflen=2097152\r\nfio: io_u error on file \\\\.\\PhysicalDrive1: Input/output error: write offset=530579456, buflen=2097152\r\n'.
```

**Comment by petrutlucian94** [comment]
We couldn't reproduce this locally yet, so we intend to collect more data from the CI jobs:

* ceph client and cluster logs
* windows event logs
* update test_rbd_wnbd.py to aggregate latency results (atm it only shows the bandwidth data)

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by idryomov** [comment]
> Looks like one of the OSDs crashed:

I would have expected to see a `RECENT_CRASH` warning as well.  It looks like it's just not running for some reason.

> 
> ```
>   cluster:
>     id:     7899ce6b-5526-40a1-8ee5-9c92d4cf8777
>     health: HEALTH_WARN
>             Module 'rbd_support' has failed dependency: No module named 'dateutil'

This is weird but probably not related.

>             Reduced data availability: 28 pgs stale
>             Degraded data redundancy: 2/240 objects degraded (0.833%), 1 pg degraded, 1 pg undersized
>             1 pool(s) do not have an application enabled
>             3 pool(s) have no replicas configured

Is this accurate?  Does the test create four different pools, three of which are `size 1`?

>             301 slow ops, oldest one blocked for 1042 sec, daemons [osd.1,osd.2] have slow ops.

Is this reproducible?

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by petrutlucian94** [comment]
jenkins test windows

**Comment by petrutlucian94** [comment]
Interesting, it still runs out of memory and terminates the OSD even after doubling the amount of vm memory. The OSD seems to consume 23GB of memory, yet only 5GB is allocated to memstore. That might imply a leak, especially since I'm only seeing this on Quincy.

https://pastebin.com/raw/x23F6RPW

**Comment by idryomov** [comment]
Why is memstore used at all?

**Comment by idryomov** [comment]
> Interesting, it still runs out of memory and terminates the OSD even after doubling the amount of memory assigned to the vm (32 GB -> 64 GB). The OSD seems to consume 23GB of memory, yet only 5GB is allocated to memstore. That might imply a leak, especially since I'm only seeing this on Quincy.

Perhaps it rings a bell to someone in @ceph/core?  If it's a memory leak it looks to be pretty massive, was there anything like that fixed in Reef?

**Comment by petrutlucian94** [comment]
> Why is memstore used at all? I mean, this job is running pretty heavy I/O test matrix now...

It's much faster. The more IO operations we can get in a reasonable timeframe, the higher changes of uncovering potential issues (e.g. leaks, race conditions, etc). I'd say it also avoids putting unnecessary strain on disk drives.

> Perhaps this rings a bell to someone in https://github.com/orgs/ceph/teams/core? If it's a memory leak it looks to be pretty massive, was there anything like that fixed in Reef?

I see that this patch hasn't been ported to quincy yet, this could be it: https://github.com/ceph/ceph/commit/5ae5b73e76bf28edff3bf32c7805224b31c4772b

**Comment by idryomov** [comment]
> I see that this patch hasn't been ported to quincy yet, this could be it: [5ae5b73](https://github.com/ceph/ceph/commit/5ae5b73e76bf28edff3bf32c7805224b31c4772b)

Ah, that surely looks like it ;)  Can you confirm that `valgrind --leak-check=full ./bin/unittest_pglog` leaks for you?

**Comment by petrutlucian94** [comment]
> Ah, that surely looks like it ;) Can you confirm that `valgrind --leak-check=full ./bin/unittest_pglog` leaks for you?

```
==49773== 177,433 (2,688 direct, 174,745 indirect) bytes in 7 blocks are definitely lost in loss record 81 of 81
==49773==    at 0x484A2F3: operator new[](unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==49773==    by 0xC51996: MemStore::Collection::create_object() const (MemStore.cc:1823)
==49773==    by 0xC70125: MemStore::Collection::get_or_create_object(ghobject_t) (MemStore.h:128)
==49773==    by 0xC57D89: MemStore::_touch(coll_t const&, ghobject_t const&) (MemStore.cc:1050)
==49773==    by 0xC612EE: MemStore::_do_transaction(ceph::os::Transaction&) (MemStore.cc:695)
==49773==    by 0xC64491: MemStore::queue_transactions(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, std::vector<ceph::os::Transaction, std::allocator<ceph::os::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*) (MemStore.cc:663)
==49773==    by 0xAAEE99: ObjectStore::queue_transaction(boost::intrusive_ptr<ObjectStore::CollectionImpl>&, ceph::os::Transaction&&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*) (ObjectStore.h:229)
==49773==    by 0xAF39FD: PGLogMergeDupsTest::test_disk_roundtrip() (TestPGLog.cc:2474)
==49773==    by 0xAF457B: virtual thunk to PGLogMergeDupsTest::TearDown() (TestPGLog.cc:2395)
==49773==    by 0xB2109B: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2605)
==49773==    by 0xB29FCE: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2641)
==49773==    by 0xB17440: testing::Test::Run() (gtest.cc:2688)
==49773== 
==49773== LEAK SUMMARY:
==49773==    definitely lost: 5,968 bytes in 14 blocks
==49773==    indirectly lost: 183,316 bytes in 274 blocks
==49773==      possibly lost: 0 bytes in 0 blocks
==49773==    still reachable: 3,874 bytes in 13 blocks
==49773==         suppressed: 0 bytes in 0 blocks
==49773== Reachable blocks (those to which a pointer was found) are not shown.
==49773== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==49773== 
==49773== For lists of detected and suppressed errors, rerun with: -s
==49773== ERROR SUMMARY: 5461 errors from 150 contexts (suppressed: 0 from 0)
```

I'll backport the fix.

**Comment by petrutlucian94** [comment]
backport: https://github.com/ceph/ceph/pull/50091

**Comment by idryomov** [comment]
> quincy backport: #50091

Merged, re-triggering!

**Comment by idryomov** [comment]
jenkins test windows

**Comment by idryomov** [comment]
jenkins test api

**Comment by petrutlucian94** [comment]
The good news is that the OSDs are no longer crashing because of the leak. The live resize test fails but that's expected since this feature hasn't been backported to quincy.

We can either skip the test or backport the feature (already did it downstream since there were so many people asking for it). The backport would be trivial, although I'm not sure what's the official policy on feature backports.

File: qa/workunits/windows/run-tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+safe_exec python.exe $testRbdWnbd --test-name RbdStampTest --iterations 100
+
+# It can take a while to setup the partition (~10s), we'll use fewer iterations.
+safe_exec python.exe $testRbdWnbd --test-name RbdFsTest --iterations 4
+safe_exec python.exe $testRbdWnbd --test-name RbdFsFioTest --iterations 4
+safe_exec python.exe $testRbdWnbd --test-name RbdFsStampTest --iterations 4
+
+# The live resize feature is not available in Quincy. Uncomment the following if
+# it gets backported.
+#
```
Comment (idryomov) [inline qa/workunits/windows/run-tests.ps1:30]:
```suggestion
```

**Review by idryomov — COMMENTED** [review]
Please fold the two latency reporting-related into the bulk backport commit or, if you would rather keep them, add "cherry picked from commit" tags (redo with `git cherry-pick -x`).

**Comment by idryomov** [comment]
ceph API tests failed because tracker.ceph.com is unavailable.

**Comment by idryomov** [comment]
jenkins test make check

**Review by idryomov — APPROVED** [review]

**Comment by petrutlucian94** [comment]
I saw the ``needs-qa`` label but that's probably unnecessary. This change only touches the Python rbd-wnbd tests, which have already been executed by the Jenkins Windows job.

**Comment by idryomov** [comment]
jenkins test api

**Comment by idryomov** [comment]
> I saw the `needs-qa` label but that's probably unnecessary. This change only touches the Python rbd-wnbd tests, which have already been executed by the Jenkins Windows job.

True, but this is the process backport PRs.  We try to be more strict here just to avoid getting bitten but seemingly no-op changes (it happened in the past).

**Pull request merged**